File: cpp-client/deephaven/tests/filter_test.cc
/*
 * Copyright (c) 2016-2022 Deephaven Data Labs and Patent Pending
 */
#include <iostream>
#include "tests/third_party/catch.hpp"
#include "tests/test_util.h"
#include "deephaven/client/client.h"

using deephaven::client::TableHandle;

namespace deephaven::client::tests {
TEST_CASE("Filter a Table", "[filter]") {
  auto tm = TableMakerForTests::Create();
  auto table = tm.Table();

  auto import_date = table.GetStrCol("ImportDate");
  auto ticker = table.GetStrCol("Ticker");
  auto close = table.GetNumCol("Close");

  auto t1 = table.Where(
      "ImportDate == `2017-11-01` && Ticker == `AAPL` && (Close <= 120.0 || isNull(Close))");
  std::cout << t1.Stream(true) << '\n';

  auto t2 = table.Where(import_date == "2017-11-01" && ticker == "AAPL" &&
      (close <= 120.0 || close.isNull()));
  std::cout << t2.Stream(true) << '\n';

  std::vector<std::string> import_date_data = {"2017-11-01", "2017-11-01", "2017-11-01"};
  std::vector<std::string> ticker_data = {"AAPL", "AAPL", "AAPL"};
  std::vector<double> open_data = {22.1, 26.8, 31.5};
  std::vector<double> close_data = {23.5, 24.2, 26.7};
  std::vector<int64_t> vol_data = {100000, 250000, 19000};

  const TableHandle *tables[] = {&t1, &t2};
  for (const auto *t : tables) {
    CompareTable(
        *t,
        "ImportDate", import_date_data,
        "Ticker", ticker_data,
        "Open", open_data,
        "Close", close_data,
        "Volume", vol_data
    );
  }
}
}  // namespace deephaven::client::tests

File: csharp/client/DhClientTests/FilterTest.cs
using Deephaven.DeephavenClient.Interop;
using System;
using Deephaven.DeephavenClient;
using Deephaven.DeephavenClient.Utility;
using Xunit.Abstractions;

namespace Deephaven.DhClientTests;

public class FilterTest {
  private readonly ITestOutputHelper _output;

  public FilterTest(ITestOutputHelper output) {
    _output = output;
  }

  [Fact]
  public void TestFilterATable() {
    using var ctx = CommonContextForTests.Create(new ClientOptions());
    var table = ctx.TestTable;

    var t1 = table.Where(
      "ImportDate == `2017-11-01` && Ticker == `AAPL` && (Close <= 120.0 || isNull(Close))");
    _output.WriteLine(t1.ToString(true));

    var importDateData = new[] { "2017-11-01", "2017-11-01", "2017-11-01"};
    var tickerData = new []{ "AAPL", "AAPL", "AAPL"};
    var openData = new[] { 22.1, 26.8, 31.5 };
    var closeData = new[] { 23.5, 24.2, 26.7 };
    var volData = new Int64[] { 100000, 250000, 19000 };

    var tc = new TableComparer();
    tc.AddColumn("ImportDate", importDateData);
    tc.AddColumn("Ticker", tickerData);
    tc.AddColumn("Open", openData);
    tc.AddColumn("Close", closeData);
    tc.AddColumn("Volume", volData);
    tc.AssertEqualTo(t1);
  }
}
